src/main/java/edu/ucla/library/iiif/fester/handlers/PostThumbnailsHandler.java

```diff
@@ -0,0 +1,292 @@
+
+    /**
+     * Creates a handler to handle POSTs to add thumbnail URLs.
+     *
+     * @param aVertx A Vert.x instance
+     * @param aConfig A application configuration
+     * @throws IOException If there is trouble reading the HTML template files
+     */
+    public PostThumbnailsHandler(final Vertx aVertx, final JsonObject aConfig) throws IOException {
+        super(aVertx, aConfig);
+
+        final byte[] bytes = IOUtils.readBytes(getClass().getResourceAsStream("/webroot/error.html"));
+
+        myExceptionPage = new String(bytes, StandardCharsets.UTF_8);
+        myFesterizeVersion = aConfig.getString(Config.FESTERIZE_VERSION);
+        myFUAPattern = Pattern.compile("Festerize/(?<version>\\d+\\.\\d+\\.\\d+)");
+    }
+
+    /*
+     * (non-Javadoc)
+     * @see io.vertx.core.Handler#handle(java.lang.Object)
+     */
+    @Override
+    public void handle(final RoutingContext aContext) {
+        final HttpServerRequest request = aContext.request();
+        final HttpServerResponse response = aContext.response();
+        final Set<FileUpload> csvUploads = aContext.fileUploads();
+        final String festerizeUserAgent = StringUtils.trimTo(request.getHeader("User-Agent"), Constants.EMPTY);
+        final Matcher festerizeUserAgentMatcher = myFUAPattern.matcher(festerizeUserAgent);
+        final String errorMessage;
+
+        // An uploaded CSV is required
+        if (csvUploads.isEmpty()) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_037);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else if (festerizeUserAgentMatcher.matches() &&
+                   !festerizeUserAgentMatcher.group("version").equals(myFesterizeVersion)) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_147, myFesterizeVersion);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else {
+            final FileUpload csvFile = csvUploads.iterator().next();
+            final String filePath = csvFile.uploadedFileName();
+            final String fileName = csvFile.fileName();
+
+            try (CSVReader csvReader = new CSVReader(Files.newBufferedReader(Paths.get(filePath)))) {
+                final CsvParser parser = new CsvParser().parse(Paths.get(filePath));
+                final List<String[]> originalLines = csvReader.readAll();
+                final List<String[]> linesWithThumbs = ThumbnailUtils.addThumbnailColumn(originalLines);
+                final int manifestIndex = Arrays.asList(linesWithThumbs.get(0)).indexOf(CSV.MANIFEST_URL);
+                final List<Future> futures = new ArrayList<>();
+                for (int rowIndex = 1; rowIndex < linesWithThumbs.size(); rowIndex++) {
+                    final ObjectType rowType = CsvParser.getObjectType(
+                                               linesWithThumbs.get(rowIndex), parser.getCsvHeaders());
+                    if (rowType.equals(ObjectType.WORK) || rowType.equals(ObjectType.PAGE)) {
```

Comment:
Are we doing this for pages as well as works? I think only works will have a manifest associated with them. @ksclarke ?

Reply:
Yes only works have manifests. Pages are all contained in works

Reply:
Removed PAGE clause